### ISSUE TITLE
Error calling curried `forAll`

### ISSUE THREAD

**Issue by DanielAsher** [typelift/SwiftCheck#107]
```
property("curried forAll") <- 
    forAll(Gen.pure("a"))(Gen.pure("b")) { (lhs: String, rhs: String) in 
        return lhs != rhs
    }
```
causes this error:
`Cannot invoke 'forAll' with an argument list of type '(Gen<String>, (String, String) -> Bool)'`

Am I calling this correctly? - I couldn't find any examples in `SwiftCheck`

**Comment by DanielAsher** [comment]
I also get a compiler error with the follow uncurried call to `forAll`
```
        property("Cannot invoke forAll with an argument list of type (Gen<[String]>, ([String]) -> Bool") <- 
            forAll(Gen.pure("0").proliferate()) { (zs: [String]) in 
                return false
            }  
```

**Comment by DanielAsher** [comment]
I managed to get the following working using my implementation of `TupleOf` #106 :
```
        let pairGen  = Gen<(String, String)>.zip(Gen.pure("0"), Gen.pure("1"))
        let pairsGen = pairGen.fmap { TupleOf($0) }.proliferate().fmap { x in ArrayOf(x) }
        
        property("`0` and `1` are distinct") <- forAll(pairsGen) { (pairs: ArrayOf<TupleOf<String, String>>) in
            let areDistinct = pairs.getArray.reduce(true) { (acc, t) in
                let (lhs, rhs) = t.getTuple
                return acc && lhs != rhs
            }
            return areDistinct
        }
```

**Comment by CodaFi** [comment]
Damn that's some terrible overload resolution right there.  I've fixed a lot of it in #105, but there's also some hard limitations of the language/framework you're up against.

According to head, your first example will compile now with some small changes:

```swift
property("curried forAll") <- forAll(Gen.pure("a"), Gen.pure("b")) { (lhs, rhs) in
	return lhs != rhs
}
```

The second doesn't work because `[String]` does not (and cannot) conform to `Arbitrary`.  You need to generate `ArrayOf<String>` instead which is easy enough:

```swift
property("curried forAll") <- forAll(Gen.pure("0").proliferate().fmap(ArrayOf.init)) { (zs: ArrayOf<String>) in
	return false
}
```

In the final case, just generate 2 arrays and zip them together in the testing block:

```swift
let lPair = Gen.pure("0").proliferate().fmap(ArrayOf.init)
let rPair = Gen.pure("1").proliferate().fmap(ArrayOf.init)

property("`0` and `1` are distinct") <- forAll(lPair, rPair) { (ls: ArrayOf<String>, rs: ArrayOf<String>) in
	let pairs = zip(ls.getArray, rs.getArray)
	let areDistinct = pairs.reduce(true) { (acc, t) in
		let (lhs, rhs) = t
		return acc && lhs != rhs
	}
	return areDistinct
}
```

**Comment by CodaFi** [comment]
I think I should include some non-arbitrary overloads with explicit generators (`forAllNoShrink`) for exactly the case you have here.

**Comment by CodaFi** [comment]
With #108 these can become

```swift
property("Cannot invoke forAll with an argument list of type (Gen<[String]>, ([String]) -> Bool") <- forAllNoShrink(Gen.pure("0").proliferate()) { (zs: [String]) in
	return false
}
```

and

```swift
let pairGen  = Gen<(String, String)>.zip(Gen.pure("0"), Gen.pure("1")).proliferate()

property("`0` and `1` are distinct") <- forAllNoShrink(pairGen) { (pairs: Array<(String, String)>) in
	let areDistinct = pairs.reduce(true) { (acc, t) in
		let (lhs, rhs) = t
		return acc && lhs != rhs
	}
	return areDistinct
}
```

**Comment by CodaFi** [comment]
Fixed by the merge of #108 and the comments above.

**Comment by DanielAsher** [comment]
@CodaFi thanks so much for this - your solution is far more elegant! :+1:
